src/workflow_provider/default_vm_workflow_provider.cpp

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (C) 2021 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <multipass/default_vm_workflow_provider.h>
+#include <multipass/exceptions/download_exception.h>
+#include <multipass/exceptions/invalid_memory_size_exception.h>
+#include <multipass/exceptions/workflow_exceptions.h>
+#include <multipass/format.h>
+#include <multipass/logging/log.h>
+#include <multipass/query.h>
+#include <multipass/url_downloader.h>
+#include <multipass/utils.h>
+
+#include <yaml-cpp/yaml.h>
+
+#include <QFile>
+#include <QFileInfo>
+
+#include <Poco/StreamCopier.h>
+#include <Poco/Zip/ZipArchive.h>
+#include <Poco/Zip/ZipStream.h>
+
+#include <fstream>
+#include <sstream>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+
+namespace
+{
+const QString github_workflows_archive_name{"multipass-workflows.zip"};
+const QString workflow_dir_version{"v1"};
+constexpr auto category = "workflow provider";
+
+auto workflows_map_for(const std::string& archive_file_path)
+{
+    std::map<std::string, std::string> workflows_map;
+    std::ifstream zip_stream{archive_file_path, std::ios::binary};
+    Poco::Zip::ZipArchive zip_archive{zip_stream};
+
+    for (auto it = zip_archive.headerBegin(); it != zip_archive.headerEnd(); ++it)
+    {
+        if (it->second.isFile())
+        {
+            auto file_name = it->second.getFileName();
+            QFileInfo file_info{QString::fromStdString(file_name)};
+
+            if (file_info.path().contains(workflow_dir_version) && file_info.suffix() == "yaml")
+            {
+                Poco::Zip::ZipInputStream zip_input_stream{zip_stream, it->second};
+                std::ostringstream out(std::ios::binary);
+                Poco::StreamCopier::copyStream(zip_input_stream, out);
+                workflows_map[file_info.baseName().toStdString()] = out.str();
```

Comment:
Should we do any validation on the basename?

Reply:
I'm not sure I follow.  Like what would we be validating?

Reply:
Basically it should be a valid hostname, since we want to use it as the instance name (and by extension, its hostname).

Reply:
Ah, ok, got it and yes, will do that.

Reply:
Looking at this some more, I think it would be better to validate the key name we use under `instances` in the yaml rather than the file's base name.  We could allow a bit more flexibility in the file name, but the instance name in the yaml will need to conform to a valid host name.  WDYT?

Reply:
I don't think it makes much sense to give that flexibility. If you `multipass launch minikube`, and we want the hostnames to be `minikube[-$suffix]`, no point introducing more names.

Reply:
Ok, after a brief chat, I understand what we need to do here.  Thanks!

Reply:
Ok, file name-> host name validation and using the workflow name as the instance name by default is done.